File: content/snippets/go/interview_lcci_03_03_stack_of_plates.go
package main

import "fmt"

type StackOfPlates struct {
	cap    int
	stacks [][]int
}

func NewStackOfPlates(capacity int) *StackOfPlates {
	return &StackOfPlates{cap: capacity}
}

func (s *StackOfPlates) Push(val int) {
	if s.cap <= 0 {
		return
	}
	if len(s.stacks) == 0 || len(s.stacks[len(s.stacks)-1]) == s.cap {
		s.stacks = append(s.stacks, []int{})
	}
	last := len(s.stacks) - 1
	s.stacks[last] = append(s.stacks[last], val)
}

func (s *StackOfPlates) Pop() int {
	return s.PopAt(len(s.stacks) - 1)
}

func (s *StackOfPlates) PopAt(index int) int {
	if index < 0 || index >= len(s.stacks) || len(s.stacks[index]) == 0 {
		return -1
	}
	last := len(s.stacks[index]) - 1
	val := s.stacks[index][last]
	s.stacks[index] = s.stacks[index][:last]
	if len(s.stacks[index]) == 0 {
		s.stacks = append(s.stacks[:index], s.stacks[index+1:]...)
	}
	return val
}

func main() {
	s := NewStackOfPlates(2)
	s.Push(1)
	s.Push(2)
	s.Push(3)
	s.Push(4)
	fmt.Println(s.PopAt(0))
	fmt.Println(s.Pop())
	fmt.Println(s.Pop())
	fmt.Println(s.Pop())
}

File: content/snippets/scala3/interview_lcci_03_03_stack_of_plates.scala
object Lcci0303StackOfPlates:
  final class StackOfPlates(capacity: Int) {
    private val stacks = scala.collection.mutable.ArrayBuffer.empty[scala.collection.mutable.ArrayBuffer[Int]]

    def push(value: Int): Unit =
      if (capacity <= 0) return
      if (stacks.isEmpty || stacks.last.length == capacity) {
        stacks += scala.collection.mutable.ArrayBuffer.empty[Int]
      }
      stacks.last += value

    def pop(): Int = popAt(stacks.length - 1)

    def popAt(index: Int): Int =
      if (index < 0 || index >= stacks.length || stacks(index).isEmpty) return -1
      val value = stacks(index).remove(stacks(index).length - 1)
      if (stacks(index).isEmpty) stacks.remove(index)
      value
  }

  @main def main(): Unit =
    val s = new StackOfPlates(2)
    s.push(1)
    s.push(2)
    s.push(3)
    s.push(4)
    println(s.popAt(0))
    println(s.pop())
    println(s.pop())
    println(s.pop())
